File: js/packages/teams-ai/src/prompts/PromptFunctions.ts
/**
 * @module teams-ai
 */
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */

import { TurnContext } from 'botbuilder';
import { Tokenizer } from '../tokenizers';
import { Memory } from '../MemoryFork';

/**
 * A collection of functions that can be called from a prompt template string.
 */
export interface PromptFunctions {
    /**
     * Returns true if the given function is defined.
     * @param name Name of the function to lookup.
     */
    hasFunction(name: string): boolean;

    /**
     * Looks up the given function.
     * @remarks
     * Throws an error if the function is not defined.
     * @param name Name of the function to lookup.
     */
    getFunction(name: string): PromptFunction;

    /**
     * Calls the given function.
     * @remarks
     * Throws an error if the function is not defined.
     * @param name Name of the function to invoke.
     * @param context Context for the current turn of conversation.
     * @param memory Interface used to access state variables.
     * @param tokenizer Tokenizer used to encode/decode strings.
     * @param args Arguments to pass to the function as an array of strings.
     */
    invokeFunction(
        name: string,
        context: TurnContext,
        memory: Memory,
        tokenizer: Tokenizer,
        args: string[]
    ): Promise<any>;
}

/**
 * A function that can be called from a prompt template string.
 * @param context Context for the current turn of conversation.
 * @param memory Interface used to access state variables.
 * @param functions Collection of functions that can be called from a prompt template string.
 * @param tokenizer Tokenizer used to encode/decode strings.
 * @param args Arguments to the function as an array of strings.
 */
export type PromptFunction = (
    context: TurnContext,
    memory: Memory,
    functions: PromptFunctions,
    tokenizer: Tokenizer,
    args: string[]
) => Promise<any>;

File: python/packages/ai/teams/ai/promptsv2/prompt_functions.py
"""
Copyright (c) Microsoft Corporation. All rights reserved.
Licensed under the MIT License.
"""

from abc import ABC, abstractmethod
from typing import Any, Awaitable, Callable, List

from botbuilder.core import TurnContext

from ...state import Memory
from ..tokenizers import Tokenizer

PromptFunction = Callable[
    [TurnContext, Memory, "PromptFunctions", Tokenizer, List[str]], Awaitable[Any]
]
"""
A function that can be called from a prompt template string.

Parameters:
    context (TurnContext): Context for the current turn of conversation.
    memory (Memory): Interface used to access state variables.
    functions (PromptFunctions): Collection of functions that can be called
      from a prompt template string.
    tokenizer (Tokenizer): Tokenizer used to encode/decode strings.
    args (List[str]): Arguments to the function as an array of strings.

Returns:
    Any: The return type is not specified, so it can be any type.
"""


class PromptFunctions(ABC):
    """A collection of functions that can be called from a prompt template string."""

    @abstractmethod
    def has_function(self, name: str) -> bool:
        """
        Checks if a function is defined.

        Args:
            name (str): Name of the function to lookup.

        Returns:
            bool: True if the function is defined, False otherwise.
        """

    @abstractmethod
    def get_function(self, name: str) -> PromptFunction:
        """
        Looks up a given function.

        Args:
            name (str): Name of the function to lookup.

        Raises:
            Error: If the function is not defined.

        Returns:
            PromptFunction: The function associated with the given name.
        """

    @abstractmethod
    def invoke_function(
        self, name: str, context: TurnContext, memory: Memory, tokenizer: Tokenizer, args: List[str]
    ):
        """
        Calls the given function.

        Args:
            name (str): Name of the function to invoke.
            context (TurnContext): Context for the current turn of conversation.
            memory (Memory): Interface used to access state variables.
            tokenizer (Tokenizer): Tokenizer used to encode/decode strings.
            args (List[str]): Arguments to pass to the function as an array of strings.

        Raises:
            Error: If the function is not defined.
        """
